### PR TITLE
[Authority + Adapter] Add package caching, and remove redundant digest checks + clones

### PR DESCRIPTION
**Pull request by gdanezis**
Looking at the profiler and flamegraphs given the latest batching benchmark: 
* Problem: we spend a significant amount of time hashing objects as part of checking locks. However an input object with an incorrect digest will result in the transaction or certificate subsequently not finding the lock on the locks table and being rejected. Solution: Therefore I have commented out the check.
* Problem: we spend a very log time re-de-serializing packages to determine their module id, and also their function signatures. Solution: now I made a cache next to the movevm that all threads share and that caches packages and their pre-computed functions.

On my laptop I see throughput go from 35K -> 45K TPS on move batched (100) benchmarks, and nearly no gap now between native transfer and move execution.

### CODE REVIEW

**Comment by gdanezis** [comment]
> > The package cache is shared, along with the movevm that is associated with the AuthorityState:
> > https://github.com/MystenLabs/sui/blob/f8c1b5963a6352ea26e6b9dd27f75257ced6f148/sui_programmability/adapter/src/adapter.rs#L51-L60
> 
> I must be missing something important here: how is the movevm/cache shared? We create a new VM in serving each request, and the package cache is contained in that VM as well. So where exactly did we share it across requests?

Ah, I see. This has not changed, we create and re-use a shared move VM that we store in AuthorityState:
https://github.com/MystenLabs/sui/blob/40b870d5e09a920926886e2c173b60f54fb3206d/sui_core/src/authority.rs#L764-L778

This AuthorityState in turn is shared within the AuthorityServer across all threads.

**Review by lxfind — CHANGES_REQUESTED** [review]
LGTM overall but sending back to your queue regarding @huitseeker's comment.

**Comment by gdanezis** [comment]
Re-enabled the check, to shortcut the discussion about whether it is needed or not.

File: sui_core/src/authority.rs [hunk 1]
```diff
@@ -153,7 +153,13 @@ impl AuthorityState {
                     }
                 );
 
-                // Check the digest matches
+                /*
+
+                // We will perform this check when we lookup the
+                // lock in the database. At this point if there is
+                // no match this will fail. No need to recompute
+                // hashes here.
```
[the remark this answers is not in the document]
Comment (gdanezis) [reply sui_core/src/authority.rs]: Re enabled the check. We can afford to hash more.

**Comment by gdanezis** [comment]
Many thanks for the review @lxfind and @huitseeker !

**Pull request merged**